Implemented Danswer versioning system.

ShibeTemple opened the pull request:
There are now two internal versions, one for the API server and another for the Web server.

- Web server version is stored in `/web/package.json` as was previously defined but not used. 
`"version": "0.1.0"`
https://github.com/ShibeTemple/danswer/blob/main/web/package.json.
Now accessible internally through `import { getWebVersion, getBackendVersion } from "@/lib/version"`

- API server version stored in `/backend/danswer/configs/constants.py`
`BACKEND_VERSION = "1.0.0"`
https://github.com/ShibeTemple/danswer/blob/main/backend/danswer/configs/constants.py
Accessible internally as `from danswer.configs.constants import BACKEND_VERSION`.

Both are displayed to the user on the login page and throughout the main application in the header. Denoted by **w** for web and **b** for backend followed by the corresponding version. 

The style is of course preliminary and can be changed along with the actual internal version numbers.

<img width="1232" alt="SCR-20231028-smus" src="https://github.com/danswer-ai/danswer/assets/6219801/e7f27c24-d977-4afe-aca0-01d6487bc06a">
<img width="1122" alt="SCR-20231028-srcy" src="https://github.com/danswer-ai/danswer/assets/6219801/3ca7a2db-1cae-4055-bcdd-53f18b868cfe">
`/api/version` is URL accessible and displays the backend API server version in JSON.
<img width="1122" alt="SCR-20231028-snaa" src="https://github.com/danswer-ai/danswer/assets/6219801/cde1b33c-77de-4e4a-b50b-088c8c68900e">

yuhongsun96 reviewed (commented):
We want to avoid manually updating the versions. The version should be provided during the build process that is triggered by the GitHub Action. The basic flow would be something as follows:

- A dev tags code (say v0.2.99)
- GitHub Action is triggered
- Action knows the version (v0.2.99)
- Action passes the version to the Dockerfile to build into the image, something like:
    - build-args: VERSION=${{ github.ref_name }}
- The container is built to have a VERSION environment variable with the correct default value of v0.2.99
- The API picks up this version and provides it to the UI container

UI container should have a similar flow to this.

Also if not triggered via tagging, the default Environment variable should be empty and the version should default to something like v0.2.dev. We don't bump the minor versions often so this will be fine.

Thanks for the good work! What do you think of this flow?

yuhongsun96 commented on backend/danswer/configs/constants.py:
```diff
@@ -1,5 +1,6 @@
 from enum import Enum
 
+BACKEND_VERSION = "1.0.0"
```
A standard is to have it in the __init__.py by the entrypoint which is main.py.

See here:
<img width="710" alt="image" src="https://github.com/danswer-ai/danswer/assets/32520769/de699c0d-1b8f-4831-b753-7ff22d1dc89c">

Be sure to have it be able to be read from env and prioritize that value so that it can be used with the build process

yuhongsun96 commented on web/src/app/auth/login/page.tsx:
```diff
@@ -80,6 +93,9 @@ const Page = async () => {
           </div>
         </div>
+        <div className="fixed bottom-4 right-4 z-50 text-slate-400 p-2">
+          VERSION w{ web_version } b{ backend_version }
```
Let move it to a new section on the bottom of the admin page

<img width="573" alt="image" src="https://github.com/danswer-ai/danswer/assets/32520769/34e1e6ec-d723-45b3-8403-e403ba1b6eaa">

yuhongsun96 replied:
If you want, you could also add some text saying "Danswer MIT Version" somewhere on the page

ShibeTemple commented:
PR revised with requested changes. 

Added system information page to admin panel with version info. Updated github workflows to include tagged version, and corresponding changes in the dockerfiles and codebases for web&backend to use env variables if present. Changed to 'dev' naming scheme if no env var is present to indicate local setup. This is done at the web & backend level. Dockerfiles will specify 0.1.0 if no tag was passed by GitHub Actions. Removed version from admin panel header.

<img width="967" alt="image" src="https://github.com/danswer-ai/danswer/assets/6219801/1ed8df21-0ce7-40f5-928c-221da1963395">
<img width="967" alt="image" src="https://github.com/danswer-ai/danswer/assets/6219801/6464d4c3-5312-426f-b7f4-98c1554c5a35">

yuhongsun96 reviewed (approved):
Good stuff! Thanks for the contribution!
